### ISSUE TITLE
Got “Error removing section” message after I tried to delete section (on dev.floodlightproject.org)

### ISSUE THREAD

**Issue by jwirfs-brock** [denverfoundation/storybase#723]
Story: http://dev.floodlightproject.org/build/7910b56fa75b41a9a6e9fc6aeb6baf44/

Description of what I did before/after seeing the error message:
1. Created story using sandbox template, entered in story information (title, summary, byline)
2. Edited “untitled section” to create a side-by-side layout section (while doing this, I added an asset, toggled the layouts, then dragged the asset back from the asset drawer) and renamed it, “A story in only six words? No way...”
3. Created a new section (“You don’t have to be Hemingway...”)
4. Moved “You don’t have to be Hemingway...” section in front of “A story in only six words? No way...”
5. Added a text asset to “You don’t have to be Hemingway...”
6. Attempted to delete “You don’t have to be Hemingway...” (from the view at that section)
7. Saw “Error removing section” at top of builder (also green and blue messages -- I am not sure what they were). But now the view moved back to the story information page.
8. Observed that the section is still in the table of contents.
9. Tried deleting it again in the table of contents. This time got, “You must have at least one section” error.
10. Clicking on “You don’t have to be Hemingway...” in table of contents: section looks blank (isn’t actually there, but is still showing up in TOC).
    11.After that, I tried adding a new section. The order of sections is garbled (see screenshot below).
    12.When I exited the builder and re-entered, it appeared with ALL the sections (including the one I attempted to delete) displayed in the proper order.

Note: I have not been able to reproduce it. Am trying again...

![Screen Shot 2013-04-17 at 12 48 26 PM](https://f.cloud.github.com/assets/1020518/392656/e3f9b63e-a783-11e2-946d-ea5d56fbc8a6.png)

**Comment by ghing** [comment]
I was able to reproduce this on dev, but not on my local machine.  The DELETE request to `/api/0.1/stories/STORY_ID/sections/SECTION_ID/`, where that's the appropriate STORY_ID and SECTION_ID for the "You don't have to be Hemingway" section failed with a 404 error. 

Strangely, a get request to that same endpoint returned the section information.

I need to figure out what's causing this.

**Comment by ghing** [comment]
The `DoesNotExist` exception is thrown by the call to `obj.delete` in [storybase.api.resources](https://github.com/PitonFoundation/atlas/blob/97169fdd770b0fd3f34fe96389dccf788878bfa8/apps/storybase/api/resources.py#L297), which triggers the `storybase_story.search_indexes.StoryIndex.asset_relation_update_object` signal handler.

Here's the traceback:

```
Traceback (most recent call last):
  File "/srv/www/atlas_dev/atlas/apps/storybase/api/resources.py", line 306, in obj_delete
    obj.delete()
  File "/srv/www/atlas_dev/venv/lib/python2.7/site-packages/django/db/models/base.py", line 576, in delete
    collector.delete()
  File "/srv/www/atlas_dev/venv/lib/python2.7/site-packages/django/db/models/deletion.py", line 61, in decorated
    func(self, *args, **kwargs)
  File "/srv/www/atlas_dev/venv/lib/python2.7/site-packages/django/db/models/deletion.py", line 269, in delete
    sender=model, instance=obj, using=self.using
  File "/srv/www/atlas_dev/venv/lib/python2.7/site-packages/django/dispatch/dispatcher.py", line 172, in send
    response = receiver(signal=self, sender=sender, **named)
  File "/srv/www/atlas_dev/atlas/apps/storybase_story/search_indexes.py", line 169, in asset_relation_update_object
    self.update_object(instance.section.story)
  File "/srv/www/atlas_dev/venv/lib/python2.7/site-packages/django/db/models/fields/related.py", line 350, in __get__
    rel_obj = qs.get(**params)
  File "/srv/www/atlas_dev/venv/lib/python2.7/site-packages/django/db/models/query.py", line 366, in get
    % self.model._meta.object_name)
DoesNotExist: Section matching query does not exist.
```

I think this is a race condition because `StoryIndex.asset_relation_update_object` is trying to access the section instance (via `instance.section`) after it has already been deleted.

**Comment by ghing** [comment]
@jwirfs-brock Turns out you found a weird edge-case unrelated to the refactor.  It occurs when deleting a section containing a text asset.  You can work around it by removing the assets before deleting the section.  I need to do some thinking about the best way to fix this.

**Comment by ghing** [comment]
Fixed this in 9639bba. Closing.
